addons/hr/models/hr_plan.py

```diff
@@ -59,6 +82,142 @@ class HrPlan(models.Model):
+        if self.plan_type == 'onboarding' and self.trigger:
+            return self.trigger
+        return 'manual'
+
+    def _default_trigger_offboarding(self):
+        if self.plan_type == 'offboarding' and self.trigger:
+            return self.trigger
+        return 'manual'
+
     name = fields.Char('Name', required=True)
     plan_activity_type_ids = fields.Many2many('hr.plan.activity.type', string='Activities')
     active = fields.Boolean(default=True)
+    plan_type = fields.Selection(
+        [
+            ('onboarding', 'Onboarding'),
+            ('offboarding', 'Offboarding'),
+            ('other', 'Other'),
+        ],
+        string='Type',
+        default='onboarding',
+        required=True,
+    )
+    trigger_onboarding = fields.Selection(
+        [
+            ('manual', 'Manual'),
+            ('employee_creation', 'Employee Creation'),
+        ],
+        default=_default_trigger_onboarding,
+        required=True,
+        store=False,
+    )
+    trigger_offboarding = fields.Selection(
+        [
+            ('manual', 'Manual'),
+            ('employee_archive', 'Employee Archived'),
+        ],
+        default=_default_trigger_offboarding,
+        required=True,
+        store=False,
+    )
+    trigger_other = fields.Selection(
+        [
+            ('manual', 'Manual'),
+        ],
+        default='manual',
+        required=True,
+        store=False,
+    )
+    trigger = fields.Char(compute='_compute_trigger', store=True)
+    company_id = fields.Many2one(
+         'res.company',
+         string='Company',
+         default=lambda self: self.env.company,
+    )
+
+    @api.depends('plan_type', 'trigger_onboarding', 'trigger_offboarding', 'trigger_other')
+    def _compute_trigger(self):
+        '''
+        Get the trigger from any trigger options using the plan_type
+        '''
+
+        type_to_trigger = {
+            'onboarding': 'trigger_onboarding',
+            'offboarding': 'trigger_offboarding',
+            'other': 'trigger_other',
+        }
+        for record in self:
+            record.trigger = record[type_to_trigger[record.plan_type]]
+
+    @api.model
+    def _launch_plans(self, trigger, employees):
+        '''
+        Launches all plans for given trigger for all employees depending on their company_id
+
+        Returns False if no plans are launched, True otherwise
+        '''
+        if not trigger or not employees:
+            return False
+        company_ids = employees.mapped('company_id.id')
+        hr_plans = self.search(
+            [
+                '&',
+                ('trigger', '=', trigger),
+                '|',
+                ('company_id', '=', False),
+                ('company_id', 'in', company_ids),
+            ],
+        )
+        if not hr_plans:
+            return False
+        # Filter once per company
+        plans_per_company = dict((id, hr_plans.filtered(lambda p: p.company_id.id == int(id))) for id in company_ids)
```

Comment:
To avoid double loop, and O(n*m):
```suggestion
        plans_per_company = defaultdict(lambda: self.env['hr.plan'])
        for plan in hr_plans:
            plans_per_company[plan.company_id.id] |= plan
```

Reply:
Nice